src/descriptor/mod.rs

```diff
@@ -871,6 +942,61 @@ impl Descriptor<DescriptorPublicKey> {
         )
         .expect("Translation fn can't fail.")
     }
+
+    /// Parse a descriptor that may contain secret keys
+    ///
+    /// Internally turns every secret key found into the corresponding public key and then returns a
+    /// a descriptor that only contains public keys and a map to lookup the secret key given a public key.
+    pub fn parse_secret(s: &str) -> Result<(Descriptor<DescriptorPublicKey>, KeyMap), Error> {
+        fn parse_key(
+            s: &String,
+            keymap: &mut KeyMap,
+        ) -> Result<DescriptorPublicKey, DescriptorKeyParseError> {
+            let (public_key, secret_key) = match DescriptorSecretKey::from_str(s) {
+                Ok(sk) => (sk.as_public()?, Some(sk)),
+                Err(_) => (DescriptorPublicKey::from_str(s)?, None),
+            };
+
+            if let Some(secret_key) = secret_key {
+                keymap.insert(public_key.clone(), secret_key);
+            }
+
+            Ok(public_key)
+        }
+
+        let mut keymap_pk = KeyMap::new();
+        let mut keymap_pkh = KeyMap::new();
+
+        let descriptor = Descriptor::<String>::from_str(s)?;
+        let descriptor = descriptor
+            .translate_pk(
+                |pk| parse_key(pk, &mut keymap_pk),
+                |pkh| parse_key(pkh, &mut keymap_pkh),
+            )
+            .map_err(|e| Error::Unexpected(e.to_string()))?;
+
+        keymap_pk.extend(keymap_pkh.into_iter());
+
+        Ok((descriptor, keymap_pk))
+    }
+
+    pub fn to_string_with_secret(&self, key_map: &KeyMap) -> String {
```

Comment:
This could also be done with far fewer allocations, but we can leave it to a followup PR